### PR TITLE
fix: only disallow `xml` by itself in PITarget

### PR DESCRIPTION
**Pull request by amaanq**

### CODE REVIEW

**Comment by ObserverOfTime** [comment]
Unfortunately this doesn't solve it.

```bash
tree-sitter parse --scope source.xml /dev/stdin <<'EOF'
<?xmlfoo bar?>
<foo/>
EOF
```
```scheme
(document [0, 0] - [2, 0]
  (ERROR [0, 0] - [1, 0]
    (Name [0, 2] - [0, 8])
    (Name [0, 9] - [0, 12]))
  root: (element [1, 0] - [1, 6]
    (EmptyElemTag [1, 0] - [1, 6]
      (Name [1, 1] - [1, 4]))))
 ```

**Comment by amaanq** [comment]
ah, so xml at the start is fine if the next char is `isalpha`?

**Comment by ObserverOfTime** [comment]
It can contain 'xml' but it cannot be *just* 'xml'.
The [standard](https://www.w3.org/TR/xml11/#NT-PITarget) is not entirely clear about it but that's how the LSP works.

**Comment by amaanq** [comment]
Is `<?xml foobar?>` allowed? I don't know how/where I can check that

**Comment by ObserverOfTime** [comment]
It's not. But `scan_pi_target` should only be scanning the `xml` part.

**Comment by amaanq** [comment]
Hm, from this https://www.w3schools.com/xml/xml_validator.asp any alnum, -, or _ following xml is ok

**Comment by ObserverOfTime** [comment]
Here's an idea I had that is not quite working:

```c
/// Check if the lookahead can appear in PITarget
static inline bool match_pi_target(int32_t lookahead) {
    return isalnum(lookahead) || lookahead == '_' || lookahead == ':' ||
           lookahead == '.' || lookahead == '-' || lookahead == 183;
}

/// Scan for the target of a PI node
static bool scan_pi_target(TSLexer *lexer) {
    short match = 0;

    if (isalpha(lexer->lookahead) || lexer->lookahead == '_') {
        if (lexer->lookahead == 'x' || lexer->lookahead == 'X') ++match;
        advance(lexer);
    } else {
        return false;
    }

    while (match_pi_target(lexer->lookahead)) {
        switch (match) {
            case 1:
                match = (lexer->lookahead == 'm' || lexer->lookahead == 'M') ? 2 : 0;
                break;
            case 2:
                match = (lexer->lookahead == 'l' || lexer->lookahead == 'L') ? 3 : 0;
                break;
            case 3:
                if (!match_pi_target(lexer->lookahead)) return false;
                break;
            default:
                break;
        }
        lexer->mark_end(lexer);
        advance(lexer);
    }

    lexer->result_symbol = PITarget;
    return true;
}
```

**Comment by amaanq** [comment]
Think I got it

**Comment by amaanq** [comment]
Note that your example doesn't seem to be a valid xml file, but now xmlfoo gets recognized as a PITarget

image

**Comment by ObserverOfTime** [comment]
> Note that your example doesn't seem to be a valid xml file

But it should be. :thinking:

File: common/scanner.h [hunk 1]
```diff
@@ -27,27 +32,37 @@ static bool scan_pi_target(TSLexer *lexer) {
...
                 advance(lexer);
                 if (lexer->lookahead == 'l' || lexer->lookahead == 'L') {
                     advance(lexer);
-                    return false;
+                    if (is_valid_pi_char(lexer->lookahead)) {
+                        found_x_first = false;
+                        bool last_char_hyphen = lexer->lookahead == '-';
+                        advance(lexer);
+                        if (last_char_hyphen) {
+                            // scan for stylesheet/model and disallow that
```
Comment (ObserverOfTime) [inline common/scanner.h:46]: It's not actually an error to allow those.

Comment (amaanq) [reply common/scanner.h:46]: it was overriding them in the tests

Comment (ObserverOfTime) [reply common/scanner.h:46]: Does increasing the precedence not work?

Comment (amaanq) [reply common/scanner.h:46]: throwing the string literal as an external without processing it and checking if it's valid could work

Comment (ObserverOfTime) [reply common/scanner.h:48]: There has to be a better solution...

Comment (amaanq) [reply common/scanner.h:48]: There really isn't :grimacing:, either xml-model in the scanner will override the literal string unconditionally or we disallow it at all or we check if the string is valid and then disallow it if it is valid

**Comment by ObserverOfTime** [comment]
> Note that your example doesn't seem to be a valid xml file

It seems that `scan_pi_content` is not working as intended.
`<?xmlfoo?>` is parsed properly but `<?xmlfoo bar?>` is not.

**Comment by amaanq** [comment]
Ah I see, when we check for ?> we are actually just done with that token, but we return false when we shouldn't - thanks for the catch

**Pull request merged**